### PR TITLE
return fake create query when executing `show create table` on system's tables.

### PR DESCRIPTION
**Pull request by cmsxbc**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
return  fake create query when executing `show create table` on system's tables.

Detailed description / Documentation draft:
Current exception thrown when executing `SHOW CREATE TABLE system.users` is `Cannot open file`. It confused a lot of users that first time tried to get structure of system tables. Such as #29283, and there is no document mention it.
So this pr contains:
1. Throw exactly exception that some tables cannot be the target of `SHOW CREATE TABLE`.
2. Mention it in the document.

After discuss,  decided to show a fake query for these tables.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Ok.

What is confusing me is that getting CREATE TABLE statement is the responsibility of IDatabase, not IStorage.
So, the method `supportsShowCreateTable` in IStorage is confusing. Let's ask @tavplubix about ideas...

**Comment by cmsxbc** [comment]
@alexey-milovidov Because `system` is just a normal `DatabaseOnDisk`  and user can create any table under `system`. So we shall not disable it on IDatabase.
By the way, I will bypass style check because `SHOW CREATE TABLE` cannot use `database = currentDatabase()/$CLICKHOUSE_DATABASE` condition, as the check expected [check style result](https://clickhouse-test-reports.s3.yandex.net/31391/49abeb14b4dab8dade5a9bdae5a3e1a980f58f4e/style_check/style_output.txt)

**Comment by tavplubix** [comment]
Let's rename `supportsShowCreateTable()` to `isSystemStorage()` or something like that.

**Review by tavplubix — APPROVED** [review]

**Comment by vitlibar** [comment]
I think maybe `SHOW CREATE TABLE` can show some fake `CREATE TABLE` statement for system tables, for example:

```
:) SHOW CREATE TABLE system.one

CREATE TABLE system.one (dummy UInt8) ENGINE=System
```

What do you think?

**Comment by tavplubix** [comment]
> I think maybe `SHOW CREATE TABLE` can show some fake `CREATE TABLE` statement for system tables, for example:
> 
> ```
> :) SHOW CREATE TABLE system.one
> 
> CREATE TABLE system.one (dummy UInt8) ENGINE=System
> ```
> 
> What do you think?

It's good idea, we can create `ASTColumns` from `ColumnsDescription`, take engine name from `IStorage::getName()` and manually construct create query for system table. But we still need something like `IStorage::isSystemStorage()` method for this. Or we can do it for all tables and always show "fake" create query if metadata file does not exist, but it's questionable, because such behavior may hide some errors or bugs (but maybe we can add some assertions).

**Comment by cmsxbc** [comment]
I think it's not a good idea to do it for all tables.The result of `SHOW CREATE TABLE` should be reusable, I always get the create query from one server and apply it to another server. It's confused if the statement from `SHOW CREATE TABLE` cannot be executed, unless I know exactly the statement is fake.

**Comment by vitlibar** [comment]
> I think it's not a good idea to do it for all tables.The result of `SHOW CREATE TABLE` should be reusable, I always get the create query from one server and apply it to another server. It's confused if the statement from `SHOW CREATE TABLE` cannot be executed, unless I know exactly the statement is fake.

You will able to understand that the table is internal from the words `ENGINE=System` in its definition and knowing that the table engine `System` cannot be used to create tables. Or we could make the fact that the table is system more prominent by starting our "fake create statement" with `CREATE SYSTEM TABLE ...` if `ENGINE=System` doesn't seem to be visible enough. My point is that some people might use `SHOW CREATE TABLE` to get the information about the columns, their names and types, and not to recreate the table.

**Comment by cmsxbc** [comment]
@vitlibar @tavplubix I have implemented one

File: src/Databases/DatabaseOnDisk.cpp [hunk 1]
```diff
@@ -673,6 +679,69 @@ ASTPtr DatabaseOnDisk::getCreateQueryFromMetadata(const String & database_metada
...
+                Expected expected;
+                expected.max_parsed_pos = string_end;
+                Tokens tokens(type_name.c_str(), string_end);
+                IParser::Pos pos(tokens, getContext()->getSettingsRef().max_parser_depth);
+                ParserDataType parser;
+                if (!parser.parse(pos, ast_type, expected))
+                {
+                    /// this should never occur, shall we change to use assert?
+                    if (throw_on_error)
+                        throw Exception(ErrorCodes::CANNOT_GET_CREATE_TABLE_QUERY, "Cannot parser metadata of {}.{}", backQuote(getDatabaseName()), backQuote(table_name));
```
Comment (tavplubix) [inline src/Databases/DatabaseOnDisk.cpp:730]:
```suggestion
                    if (throw_on_error)
                        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot parser metadata of {}.{}", backQuote(getDatabaseName()), backQuote(table_name));
```

Comment (tavplubix) [reply src/Databases/DatabaseOnDisk.cpp:730]: Exceptions with `LOGICAL_ERROR` code work like assert in debug builds and like ordinary exception in release builds

File: src/Databases/DatabaseOnDisk.cpp [hunk 2]
```diff
@@ -673,6 +679,69 @@ ASTPtr DatabaseOnDisk::getCreateQueryFromMetadata(const String & database_metada
...
+
+    /// setup create table query storage info.
+    {
+        auto ast_engine = std::make_shared<ASTFunction>();
+        ast_engine->name = storage->getName();
+        auto ast_storage = std::make_shared<ASTStorage>();
+        ast_storage->set(ast_storage->engine, ast_engine);
+        create_table_query->set(create_table_query->storage, ast_storage);
+    }
+    /// setup create table query columns info.
```
Comment (tavplubix) [inline src/Databases/DatabaseOnDisk.cpp:707]: Btw, we already have similar code [here](https://github.com/ClickHouse/ClickHouse/blob/7b5954b736465aa059803139a3870aeab3d2d327/src/Databases/MySQL/DatabaseMySQL.cpp#L116) and [here](https://github.com/ClickHouse/ClickHouse/blob/d76976e272268f76bb469d611aca66acd9dd1cbf/src/Databases/SQLite/DatabaseSQLite.cpp#L173). Is it possible to extract common code to some function or static method?

Comment (cmsxbc) [reply src/Databases/DatabaseOnDisk.cpp:707]: Ok, I will have a try.

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
Integration tests - known issue with iptables rules
Performance - #31530
Stress test (debug, actions) - #31531

**Pull request merged**